backend/logic/courses/courses.js

```diff
@@ -261,12 +261,108 @@ var updateCourse = async (course, newTitle, newTags) => {
   }
 };
 
+/**
+ * Assign a course to a curriculum
+ * @author Yufei Liu
+ * @param {string} curriculumType,{string} courseCode,{string} curriculum
+ * @returns true if successful
+ * @throws error if curriculumType, courseCode or curriculum dose not exist
+ *         course already assigned a curriculum
+ *
+ */
+var assignCourseToCurriculum = async (
+  curriculumType,
+  courseCode,
+  curriculum
+) => {
+  if (!format.isMcGillCourse(courseCode)) {
+    throw Error("Invalid course format!\n");
+  }
+
+  let conn = await mysql.getNewConnection();
+
+  const checkExistQuery = `SELECT COUNT(*) AS count
+  FROM (
+      SELECT course_code, curriculum_name FROM curriculum_tech_comps
+      union all
+      SELECT course_code, curriculum_name FROM curriculum_complementaries
+      union all
+      SELECT course_code, curriculum_name FROM curriculum_core_classes
+  ) a
+  WHERE course_code = ? AND curriculum_name = ?;`;
+
+  // Check if the course has already been assigned, verify if curriculum type, course and curriculum are existed in DB
+  let ifAssigned;
+  let checkCourse;
+  let checkCurriculum;
+
+  try {
+    ifAssigned = await conn.query(checkExistQuery, [courseCode, curriculum]);
+    checkCourse = await conn.query(
+      "SELECT COUNT(*) AS count FROM courses WHERE course_code = ?",
+      [courseCode]
+    );
+    checkCurriculum = await conn.query(
+      "SELECT COUNT(*) AS count FROM curriculums WHERE curriculum_name = ?",
+      [curriculum]
+    );
+  } catch (err) {
+    console.log(err);
+    conn.release();
+    throw Error("Internal Server Error!\n");
+  }
+
+  if (ifAssigned[0].count !== 0) {
+    throw Error(
+      `Course ${courseCode} has already been added to ${curriculum}!\n`
+    );
+  }
+  if (checkCourse[0].count === 0) {
+    throw Error(`Course ${courseCode} does not exist!\n`);
+  }
+  if (checkCurriculum[0].count === 0) {
+    throw Error(`Curriculum ${curriculum} does not exist!\n`);
+  }
+
+  const coreTable = "curriculum_core_classes";
+  const techCompTable = "curriculum_tech_comps";
+  const compTable = "curriculum_complementaries";
+  let tableType = "";
+
+  if (curriculumType === "core") {
+    tableType = coreTable;
+  } else if (curriculumType === "techComp") {
+    tableType = techCompTable;
+  } else if (curriculumType === "complementaries") {
+    tableType = compTable;
+  } else {
+    throw Error("Invalid curriculum type!\n");
+  }
+
+  const query = `INSERT INTO  ${tableType} (curriculum_name, course_code)
+                VALUE (?, ?);`;
+
+  try {
+    await conn.beginTransaction();
```

Comment:
Someone with better SQL experience can probably address this better than me but since you have a single SQL query to execute, wouldn't the transaction be unnecessary? It's not like you have multiple queries that need to be run sequentially and depend on each other.

Reply:
I see your point, however the transaction here is more like a lock to avoid race condition, since I'm doing `write` operation here.

Reply:
Ok, not totally convinced that it is necessary but it's fine the way it is. Planning to discuss this further with @si-x-li in person.